File: PC/androidusbsocket/src/main/java/com/example/PcClient.java
package com.example;

import java.io.*;
import java.net.Socket;
import java.util.Scanner;

public class PcClient {

    private static boolean mRunning = true;

    // Set up adb port forwarding
    public static void main(String[] args) {
        if (!setupAdbForward()) {
            System.out.println("Failed to set up port forwarding");
            return;
        } else {
            System.out.println("Port forwarding set up successfully"); 
        }

        // Prompt user for input
        System.out.println("Enter any string and press enter to send");
        // Start client
        startClient();
    }

    // Method to set up adb port forwarding
    private static boolean setupAdbForward() {
        try {
             // Execute adb command to forward from PC port 8000 to device port 9000
            Process process = Runtime.getRuntime().exec("adb forward tcp:8000 tcp:9000"); 
            process.waitFor();
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    // Method to start the Socket client
    private static void startClient() {
        try {
            // Create Socket connected to port 8000, which is forwarded to device port 9000
            Socket socket = new Socket("127.0.0.1", 8000);
            // Start a thread to continuously receive data from server
            new Thread(new InThread(socket)).start(); 
            // Read user input with Scanner
            Scanner scanner = new Scanner(System.in);
            while (true) {
                String msg = scanner.next();
                // Send user input string to server
                sendToServer(socket, msg);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    // Thread to continuously receive server data
    static class InThread implements Runnable {
        private Socket socket;

        InThread(Socket socket) {
            this.socket = socket;
        }

        @Override
        public void run() {
            while (mRunning) {
                if (socket.isClosed()) {
                    mRunning = false;
                    break;
                } 
                try {
                    // Receive data from server via Socket input stream
                    DataInputStream dis = new DataInputStream(socket.getInputStream());
                    byte[] buffer = new byte[256];
                    int len = dis.read(buffer);  
                    if (len > 0) {
                        // Print received data
                        System.out.println("\nReceived:" + new String(buffer, 0, len, "UTF-8"));
                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }
    }

    // Method to send data to server 
    public static void sendToServer(Socket socket, String msg) throws IOException {
        DataOutputStream dos = new DataOutputStream(socket.getOutputStream());
        dos.writeUTF(msg);
        dos.flush();
        // Add brief delay before sending next message
        try {
            Thread.sleep(1000L);  
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: PC/androidusbsocket_Py/PcClient.py
import socket
import subprocess
import time
import threading

# Define host and port (forwarded from Android port 9000)
HOST = "127.0.0.1"
PORT = 8000

def setup_adb_forward():
    try:
        # Forward PC port 8000 to device port 9000
        subprocess.run(["adb", "forward", "tcp:8000", "tcp:9000"])
        return True
    except Exception as e:
        print(e)
        return False

# Function to send data to server
def send_to_server(socket, message):
    try:
        # Send message as UTF-encoded bytes
        socket.sendall(message.encode("utf-8"))
        print(f"Sent: {message}")
    except Exception as e:
        print(f"Error sending data: {e}")

# Start client
def main():
    try:
        setup_adb_forward()
        # Create socket connection
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
            sock.connect((HOST, PORT))

            # Start background thread to receive data
            receive_thread = threading.Thread(target=receive_data, args=(sock,))
            receive_thread.daemon = True
            receive_thread.start()

            # Get user input and send messages until program exits
            while True:
                message = input("Enter message to send: ")
                send_to_server(sock, message)
                time.sleep(1)  # Delay between sending messages

    except Exception as e:
        print(f"Error: {e}")

def receive_data(socket):
    while True:
        try:
            data = socket.recv(1024).decode("utf-8")
            if not data:
                print("Connection closed")
                break

            lines = data.splitlines()
            print(f"Reiceived: {lines[-1]}")

        except Exception as e:
            print(f"Error receiving data: {e}")
            break

if __name__ == "__main__":
    main()
